qiskit/backends/ibmq/credentials/__init__.py

```diff
@@ -0,0 +1,62 @@
+# -*- coding: utf-8 -*-
+
+# Copyright 2018, IBM.
+#
+# This source code is licensed under the Apache License, Version 2.0 found in
+# the LICENSE.txt file in the root directory of this source tree.
+
+"""
+Utilities for working with credentials for the wrapper.
+"""
+from collections import OrderedDict
+import logging
+
+from qiskit import QISKitError
+from .credentials import Credentials
+from ._configrc import read_credentials_from_qiskitrc, store_credentials
+from ._environ import read_credentials_from_environ
+from ._qconfig import read_credentials_from_qconfig
+
+logger = logging.getLogger(__name__)
+
+
+def discover_credentials():
+    """
+    Automatically discover credentials for online providers.
+
+    This method looks for credentials in the following locations, in order,
+    and returning as soon as credentials are found::
+
+        1. in the `Qconfig.py` file in the current working directory.
+        2. in the environment variables.
+        3. in the `qiskitrc` configuration file.
+
+    Returns:
+        dict: dictionary with the contents of the configuration file, with
+            the form::
+
+            {'provider_name': {'token': 'TOKEN', 'url': 'URL', ... }}
+    """
+    credentials = OrderedDict()
+
+    # dict[str:function] that defines the different locations for looking for
+    # credentials, and their precedence order.
+    readers = OrderedDict([
+        ('qconfig', read_credentials_from_qconfig),
+        ('environment variables', read_credentials_from_environ),
+        ('qiskitrc', read_credentials_from_qiskitrc)
+    ])
+
+    # Attempt to read the credentials from the different sources.
+    for display_name, reader_function in readers.items():
+        try:
+            credentials = reader_function()
+            logger.info('Using credentials from %s', display_name)
+            if credentials:
+                break
+        except QISKitError as ex:
+            logger.warning(
+                'Automatic discovery of %s credentials failed: %s',
+                display_name, str(ex))
+
+    return credentials
```

Comment:
Do we want to return an empty ordered dict here if there are no matches? I feel like we should error or return an explicit None at least.By calling this function we're asking get us are credentials from the places they could live, and if there aren't any that feels like it's something to error on.

Reply:
I feel that being consistent in the return type (ie. always an `OrderedDict`, regardless of the number of elements it has, including 0) is more natural given the way this function is used. It basically guarantees that it will make an effort to find credentials and return a dict - and the caller decides whether having found no credentials merits to be treated exceptionally.